pragma-node/src/infra/repositories/onchain_repository.rs

```diff
@@ -0,0 +1,163 @@
+                ) AS MedianPrices
+            ) AS aggregated_price"
+        }
+        AggregationMode::Twap => Err(InfraError::InternalServerError)?,
+    };
+    Ok(query)
+}
+
+fn get_table_name_from_network(network: Network) -> &'static str {
+    match network {
+        Network::Testnet => "spot_entry",
+        Network::Mainnet => "mainnet_spot_entry",
+    }
+}
+
+fn build_sql_query(network: Network, aggregation_mode: AggregationMode) -> String {
+    format!(
+        r#"
+        WITH RankedEntries AS (
+            SELECT 
+                *,
+                ROW_NUMBER() OVER (PARTITION BY publisher, source ORDER BY timestamp DESC) as rn
+            FROM 
+                {table_name}
+            WHERE 
+                pair_id = $1 
+                AND timestamp BETWEEN (to_timestamp($2) - INTERVAL '{backward_interval}') AND to_timestamp($2)
+        ),
+        FilteredEntries AS (
+            SELECT *
+            FROM RankedEntries
+            WHERE rn = 1
+        ),
+        AggregatedPrice AS (
+            SELECT {aggregation_subquery}
+            FROM FilteredEntries
+        )
+        SELECT DISTINCT 
+            FE.*,
+            AP.aggregated_price
+        FROM 
+            FilteredEntries FE,
+            AggregatedPrice AP
+        ORDER BY 
+            FE.timestamp DESC;
+    "#,
+        table_name = get_table_name_from_network(network),
+        backward_interval = BACKWARD_TIMESTAMP_INTERVAL,
+        aggregation_subquery = get_aggregation_query(aggregation_mode).unwrap()
+    )
+}
+
+// TODO(akhercha): Only works for Spot entries
+pub async fn get_sources_and_aggregate(
+    pool: &Pool,
+    network: Network,
+    pair_id: String,
+    timestamp: u64,
+    aggregation_mode: AggregationMode,
+) -> Result<(BigDecimal, Vec<OnchainEntry>), InfraError> {
+    let raw_sql = build_sql_query(network, aggregation_mode);
+
+    let conn = pool.get().await.map_err(adapt_infra_error)?;
+    let raw_entries = conn
+        .interact(move |conn| {
+            diesel::sql_query(raw_sql)
+                .bind::<Text, _>(pair_id)
+                .bind::<BigInt, _>(timestamp as i64)
+                .load::<SpotEntryWithAggregatedPrice>(conn)
+        })
+        .await
+        .map_err(adapt_infra_error)?
+        .map_err(adapt_infra_error)?;
+
+    if raw_entries.is_empty() {
+        return Ok((BigDecimal::from(0), vec![]));
+    }
+
+    let aggregated_price = raw_entries.first().unwrap().aggregated_price.clone();
+    let entries: Vec<OnchainEntry> = raw_entries.into_iter().map(From::from).collect();
+    Ok((aggregated_price, entries))
+}
+
+// TODO(akhercha): Only works for Spot entries
+// TODO(akhercha): Does not give the same result than onchain call
+pub async fn get_last_updated_timestamp(
+    pool: &Pool,
+    network: Network,
+    pair_id: String,
+) -> Result<u64, InfraError> {
+    let raw_sql = format!(
+        r#"
+        SELECT 
+            *
```

Comment:
shouldn't we just select timestamp field ?

Reply:
yup true - it was to adhere to the `SpotEntry` struct but commented out this section for now, instead we take the last source timestamp as `last_updated_timestamp`